device/store.go

```
@@ -0,0 +1,113 @@
+// Copyright © 2020 Ispirata Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package device
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/astarte-platform/astarte-go/interfaces"
+)
+
+type astarteMessageInfo struct {
+	StorageId      int    `gorm:"primaryKey;autoIncrement;not null"`
+	Retention      string `gorm:"not null"`
+	AbsoluteExpiry int64  `gorm:"not null"`
+	InterfaceName  string `gorm:"not null"`
+	Path           string `gorm:"not null"`
+	Qos            uint8  `gorm:"not null"`
+	Payload        []byte `gorm:"not null"`
+}
+
+type property struct {
+	InterfaceName string `gorm:"primaryKey;not null"`
+	Path          string `gorm:"primaryKey;not null"`
+	RawValue      []byte `gorm:"not null"`
+}
+
+func (d *Device) getDbDir() string {
+	dbDir := filepath.Join(d.persistencyDir, "db")
+	os.MkdirAll(dbDir, 0700)
+	return dbDir
+}
+
+func (d *Device) migrateDb() error {
+	if err := d.db.AutoMigrate(&astarteMessageInfo{}, &property{}); err != nil {
+		fmt.Printf("Error in database migration: %s\n", err.Error())
```

I think you should return an error here